rpp30-boc-heartbleed/whimsical
# Continuous integration

## Description
Installed jest and added a [.yml](https://blog.stackpath.com/yaml/) file in `.github/workflows` that automates testing for each PR into the `dev` branch. I added this branch as another layer between development and production to better ensure that we always have a working app in production. Here's how the workflow will end up looking after we've fully implemented Continuous Deployment:

1. Make a PR from `feature-branch` into `dev`, triggering automated testing.
2. If tests fail, fix and test locally and then commit and push until automated testing passes.
3. Once all tests pass, manual code review as we've been doing.
4. On merge into `dev`, automatic deployment will be triggered, probably using [AWS CodePipeline](https://aws.amazon.com/codepipeline/).
5. On merge from `dev` into `main`, automatic deployment will be triggered into a staging/production environment, probably using AWS and Expo.

This is still very much a work in progress and definitely be learning as we go. Any insight is welcome and if anybody sees something that's wrong please don't hesitate to let me know. I've also included a couple handy diagrams of the ci/cd workflow below

![image](https://user-images.githubusercontent.com/55521671/147621269-224ed001-5a33-4ee8-beb8-90c8e843d45e.png)

![image](https://user-images.githubusercontent.com/55521671/147626268-45369808-6510-46c3-a143-9e8852519ef8.png)

## How to test
After making a Pull Request go to the Actions tab and check that your PR passed the tests.

## Notes
- I've changed the default branch for this repo to `dev`.
- The current .yml script is pretty basic. I plan on adding more, including a script that deploys and builds on expo, then publishes a qr code in the PR itself. This way when code reviewing we'll be able to scan on our phone and see the app (with the newly committed code) for ourselves.

## Issue tracking
https://trello.com/c/QK2q471t/9-continous-integration-needs-1
